### PR TITLE
Fix OnReturn secondary resource pipeline for HasOne relationships

### PR DESCRIPTION
**Pull request by maurei**
Additional fix for #924 that covers the case of HasOne relationships

### CODE REVIEW

File: test/JsonApiDotNetCoreExampleTests/Acceptance/ResourceDefinitions/ResourceDefinitionTests.cs [hunk 1]
```diff
@@ -241,6 +241,28 @@ public async Task Article_Through_Secondary_Endpoint_Is_Hidden()
...
+            var passport = new Passport(_dbContext) {IsLocked = true};
+            person.Passport = passport;
+
+            _dbContext.People.Add(person);
+            await _dbContext.SaveChangesAsync();
+
+            var route = $"/api/v1/people/{person.Id}/passport";
+
+            // Act
+            var response = await _client.GetAsync(route);
```
Comment (bart-degreed) [inline test/JsonApiDotNetCoreExampleTests/Acceptance/ResourceDefinitions/ResourceDefinitionTests.cs:258]: Assertion on http status code is missing.

File: test/JsonApiDotNetCoreExampleTests/Acceptance/ResourceDefinitions/ResourceDefinitionTests.cs [hunk 2]
```diff
@@ -241,6 +241,28 @@ public async Task Article_Through_Secondary_Endpoint_Is_Hidden()
             Assert.DoesNotContain(toBeExcluded, body);
         }
 
+        [Fact]
+        public async Task Passport_Through_Secondary_Endpoint_Is_Hidden()
+        {
+            // Arrange
+            var person = _personFaker.Generate();
+            var passport = new Passport(_dbContext) {IsLocked = true};
```
Comment (bart-degreed) [inline test/JsonApiDotNetCoreExampleTests/Acceptance/ResourceDefinitions/ResourceDefinitionTests.cs:249]: `passport` can be inlined

**Review by bart-degreed — CHANGES_REQUESTED** [review]

**Comment by woofi7** [comment]
I've pulled the branch and tested on my side with the project.

When tested it crashes on the call to `OnReturn`.

The var `resources` is of type `List<object>`  which is not `TResource`
```
Microsoft.CSharp.RuntimeBinder.RuntimeBinderException: The type 'object' cannot be used as type parameter 'TResource' in the generic type or method 'JsonApiDotNetCore.Hooks.Internal.IOnReturnHookExecutor.OnReturn<TResource>(System.Collections.Generic.IEnumerable<TResource>, JsonApiDotNetCore.Hooks.Internal.Execution.ResourcePipeline)'. There is no implicit reference conversion from 'object' to 'JsonApiDotNetCore.Resources.IIdentifiable'.
  at CallSite.Target
  at System.Dynamic.UpdateDelegates.UpdateAndExecute3
  at JsonApiDotNetCore.Hooks.Internal.ResourceHookExecutorFacade.OnReturnRelationship
    D:\workspace\JsonApiDotNetCore\src\JsonApiDotNetCore\Hooks\Internal\ResourceHookExecutorFacade.cs:131
  at JsonApiDotNetCore.Services.JsonApiResourceService`2+<GetSecondaryAsync>d__11.MoveNext
    D:\workspace\JsonApiDotNetCore\src\JsonApiDotNetCore\Services\JsonApiResourceService.cs:137
```

I have tried with this and it works:
```
if (resourceOrResources is IIdentifiable)
{
    var resources = ToList((dynamic) resourceOrResources);
    object[] array = _resourceHookExecutor.OnReturn(resources, ResourcePipeline.GetRelationship).ToArray();
    return array.SingleOrDefault();
}
```

This do not work, we need to store the result in a var and then do the SingleOrDefault.
`_resourceHookExecutor.OnReturn(resources, ResourcePipeline.GetRelationship).ToArray().SingleOrDefault()`

**Comment by woofi7** [comment]
The fix work with the first problem (image loading).

**Review by bart-degreed — APPROVED** [review]

**Pull request merged**